### DISCUSSION TITLE
MotoROS2: Fatal Error (subcode: 10)

### DISCUSSION THREAD

**Discussion by DocVaughan** [Yaskawa-Global/motoros2#292] [Troubleshooting] [answered]
First, thank you for the package. We have it running on three GP12s and a GP7. The GP12s are running beautifully. However, the GP7 frequently has a `MotoROS2: Fatal Error (subcode: 10)`. Anecdotally, it seems to be happening with more frequency over time. I've attached the `CMOS.BIN` file, the `PANELBOX.LOG`, our config file, and the log from the debug listener during controller startup.

We're on Humble and have seen the same issue on both 0.1.2 and 0.1.3. 

Thank you in advance for any guidance you can provide.

* [CMOS.zip](https://github.com/user-attachments/files/16661020/CMOS.zip)
* [PANELBOX.LOG](https://github.com/user-attachments/files/16661125/PANELBOX.LOG)
* [motoros2_config.yaml.zip](https://github.com/user-attachments/files/16661036/motoros2_config.yaml.zip)
* [20240819T095502_motoros2_debug_log.txt](https://github.com/user-attachments/files/16661188/20240819T095502_motoros2_debug_log.txt)

**Comment by gavanderhoorn** [comment]
> the GP12s are running beautifully. However, the GP7 frequently has a `MotoROS2: Fatal Error (subcode: 10)`.

do the GP12s also have additional axes? And are those also calibrated against the main robot (ie: `r1`)?

**Comment by DocVaughan** [comment]
Two of the GP12s also have a postioner. One does not. For the two with positioners, they are calibrated like the GP7.

**Comment by ted-miller** [comment, marked as answer]
> Anecdotally, it seems to be happening with more frequency over time

How frequent would you estimate? 

I have loaded your CMOS and rebooted 5 or 6 times without receiving that alarm. Given that this doesn't occur every time on your system either, my first suspect would be the networking layer. I suggest checking all connections and make sure that there aren't any IP conflicts. Additionally, double check the version of the microROS Agent on your PC. Make sure that it's the latest humble version.

--------------------------------

Also, the panelbox.log indicates that you don't have the arc-welding applications installed (which is good). But given how your robot was configured initially, I just want to emphasize that Simple Connect and UWI must not be installed alongside MotoROS2.

**Reply by gavanderhoorn** [reply]
> Given that this doesn't occur every time on your system either, my first suspect would be the networking layer

we're not actually logging the value of `ret` here:

https://github.com/Yaskawa-Global/motoros2/blob/3983f65d52d1909e2394d9161c9b14964b591cb6/src/ControllerStatusIO.c#L152-L159

I'll create a debug build with some more output to try and pinpoint the cause a bit more.

Network being flaky could indeed cause this, but without knowing what `rclc_publisher_init(..)` returns, it's difficult to say.

**Reply by gavanderhoorn** [reply]
@DocVaughan: would you perhaps also have the output of the Agent when MotoROS2 asserts?

**Reply by DocVaughan** [reply]
Thank you both for answering so quickly. 

Currently, this happens on nearly every startup, nearly immediately after connecting to the micro-ROS agent. You can see how quickly in the timestamps from the agent output (I've also attached as a txt file, in case that's more convenient for you):

```
foundries@pcwf05475:~/foundries_ws$ docker run   -it   --rm   --net=host   --user=$(id -u):$(id -g)   microros/micro-ros-agent:humble     udp4     --port 8888
[1724081261.736451] info     | UDPv4AgentLinux.cpp | init                     | running...             | port: 8888
[1724081261.736608] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 4
[1724081396.156202] info     | ProxyClient.cpp    | ProxyClient              | session hard timeout enabled | client_key: 0x7993E21B, timeout: 10000 ms
[1724081396.156250] info     | Root.cpp           | create_client            | create                 | client_key: 0x7993E21B, session_id: 0x81
[1724081396.156316] info     | SessionManager.hpp | establish_session        | session established    | client_key: 0x7993E21B, address: 192.168.1.33:26619
[1724081396.176606] info     | ProxyClient.cpp    | create_participant       | participant created    | client_key: 0x7993E21B, participant_id: 0x000(1)
[1724081397.910882] info     | ProxyClient.cpp    | create_topic             | topic created          | client_key: 0x7993E21B, topic_id: 0x000(2), participant_id: 0x000(1)
[1724081408.681763] info     | Root.cpp           | delete_client            | delete                 | client_key: 0x7993E21B
[1724081408.681807] info     | SessionManager.hpp | destroy_session          | session closed         | client_key: 0x7993E21B, address: 192.168.1.33:26619
[1724081408.681815] info     | Processor.cpp      | check_heartbeats         | Session destroyed due to liveliness timeout | client_key: 0x7993E21B, address: 192.168.1.33:26619
```

[micro_ros_agent_output.txt](https://github.com/user-attachments/files/16662387/micro_ros_agent_output.txt)

**Reply by DocVaughan** [reply]
Based on your networking hint above, we did some "normal" network debugging, swapping ethernet cables, etc. A new cable _seems_ to have helped. We're no longer crashing at startup, at least. We'll be running the system this afternoon, so I'll let you know if it's stays up and try to keep track of the frequency of the crashes, if not.

**Reply by gavanderhoorn** [reply]
@DocVaughan: I'll hold off on sharing a build with more debugging output. Let us know whether it turned out to be the cable.

**Reply by DocVaughan** [reply]
The system did stay up for the entire rest of the day (about 7 hours).

I'm sorry I was slow to update. Once we were up and running, we had time to make up. 

Thank you both again for your help and for your work on this package.

**Reply by gavanderhoorn** [reply]
Thanks for the update.

Please keep monitoring and let us know when something changes.

**Reply by gavanderhoorn** [reply]
I'll mark this as *answered*.

**Reply by DocVaughan** [reply]
Sounds good. I'm both sorry and happy that it turned out to be a simple problem/fix.

Thanks again for the help.

**Reply by gavanderhoorn** [reply]
> Sounds good. I'm both sorry and happy that it turned out to be a simple problem/fix.

no need to apologise for anything. That's what the forum is for.
